internal/queue/consumer.go

```diff
@@ -0,0 +1,286 @@
+			c.logger.Warn("job not found in database on completion", "job_id", j.ID)
+			msg.Nack(false, false) // don't requeue
+			return nil
+		}
+		msg.Nack(false, true) // requeue
+		return fmt.Errorf("failed to update job completed in database: %w", err)
+	}
+	msg.Ack(false)
+	return nil
+}
+
+// extractRaw extracts the raw payload from the AMQP message.
+func (c *AmqpConsumer) extractRaw(msg amqp.Delivery) map[string]any {
+	var payload map[string]any
+	if err := json.Unmarshal(msg.Body, &payload); err != nil {
+		c.logger.Warn("failed to extract raw payload", "error", err)
+		return nil
+	}
+	return payload
+}
+
+// consumeMsgs starts consuming messages from the AMQP queue.
+func (c *AmqpConsumer) consumeMsgs() (<-chan amqp.Delivery, error) {
+	msgs, err := c.amqpChannel.Consume(
+		c.queueName, // queue
+		"",          // consumer tag, empty string means a unique random tag will be generated
+		false,       // whether rabbitmq auto-acknowledges messages
+		false,       // whether only this consumer can access the queue
+		false,       // no-local
+		false,       // false means wait for server confirmation that consumer is registered
+		nil,         // args
+	)
+	if err != nil {
+		return nil, fmt.Errorf("failed to consume message: %w", err)
+	}
+	return msgs, nil
+}
+
+// resetConnectionOrChannelIfNecessary resets the AMQP connection or channel if they are nil or closed.
+func (c *AmqpConsumer) resetConnectionOrChannelIfNecessary() error {
+	c.mu.Lock() // Lock to prevent concurrent access to connection/channel object
+	defer c.mu.Unlock()
+
+	if c.amqpConnection == nil || c.amqpConnection.IsClosed() {
+		err := c.resetConnection()
+		if err != nil {
+			return err
+		}
+	}
+
+	if c.amqpChannel == nil || c.amqpChannel.IsClosed() {
+		err := c.resetChannel()
+		if err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+// resetConnection establishes a new AMQP connection.
+func (c *AmqpConsumer) resetConnection() error {
+	conn, err := c.connectFunc(c.uri)
+	if err != nil {
+		return fmt.Errorf("failed to connect to AMQP server: %w", err)
+	}
+	c.amqpConnection = conn
+	return nil
+}
+
+// resetChannel opens a new AMQP channel and declares the queue.
+func (c *AmqpConsumer) resetChannel() error {
+	ch, err := c.amqpConnection.Channel()
+	if err != nil {
+		return fmt.Errorf("failed to open AMQP channel: %w", err)
+	}
+	c.amqpChannel = ch
+
+	err = ch.Confirm(false)
+	if err != nil {
+		return fmt.Errorf("failed to put channel in confirm mode: %w", err)
+	}
+
+	_, err = ch.QueueDeclare(
+		c.queueName, // queueName
+		true,        // durable
+		false,       // delete when unused
+		false,       // exclusive
+		false,       // no-wait
+		nil,         // arguments
+	)
+	if err != nil {
+		return fmt.Errorf("failed to declare AMQP queue: %w", err)
+	}
+
+	return nil
+}
+
+// parseToTime parses a string pointer to time.Time, returning zero time if nil or invalid.
+func (c *AmqpConsumer) parseToTime(timeStr *string) time.Time {
```

Comment:
nit: This looks like a helper function that could be used elsewhere too. This could be converted to a static helper function by putting the logger as argument to the function and removing the requirement to be a method on AmqpConsumer